src/compiler/checker.ts

```diff
@@ -2497,6 +2498,22 @@ namespace ts {
             return false;
         }
 
+        function checkAndReportErrorForExtendingPrimitiveType(errorLocation: Node, name: __String, meaning: SymbolFlags): boolean {
+            if (meaning & (SymbolFlags.Value & ~SymbolFlags.NamespaceModule)) {
+                if (isPrimitiveTypeName(name) && !(errorLocation.parent.parent.kind & ~SyntaxKind.HeritageClause)) {
+                    const rawName = unescapeLeadingUnderscores(name);
+                    if(meaning & SymbolFlags.Interface) {
+                        error(errorLocation, Diagnostics.An_interface_cannot_extend_a_primitive_type_like_0_An_interface_can_only_extend_named_object_types, rawName);
+                    }
+                    else if(meaning & SymbolFlags.Class) {

```

Comment:
A syntactic check (of `errorLocation.parent.parent.parent` I think) is better. `meaning` is the kind of symbol *requested* by name resolution.

Also you should probably check that all those parent nodes are defined.